### PR TITLE
[PM-5500] Implement StateProvider in RouterService

### PR DESCRIPTION
**Pull request by rr-bw**
## Type of change

- Tech debt (refactoring, code cleanup, dependency upgrades, etc)

## Objective

Moves `deepLinkRedirectUrl` to RouterService.

No migration necessary since state is extremely short-lived (immediately cleared after usage). Comment if you think otherwise.

## Code changes

- `apps/web/src/app/core/router.service.ts` - add KeyDefinition and replace StateService instances with StateProvider.
- `libs/common/src/platform/abstractions/state.service.ts`, `libs/common/src/platform/services/state.service.ts` - remove from StateService 
- `libs/common/src/platform/models/domain/global-state.ts` - remove from global state

### Locked State

https://github.com/bitwarden/clients/assets/102181210/a285c1b8-1768-4c5a-aa8e-d0f91fa22927

### SSO with Trusted Devices

https://github.com/bitwarden/clients/assets/102181210/b66209d7-ca8d-41c0-9a36-f54640d0f43b

### CODE REVIEW

File: apps/web/src/app/core/router.service.ts [hunk 1]
```diff
@@ -67,14 +86,14 @@ export class RouterService {
    * @param url URL being saved to the Global State
    */
   async persistLoginRedirectUrl(url: string): Promise<void> {
-    await this.stateService.setDeepLinkRedirectUrl(url);
+    await this.deepLinkRedirectUrlState.update((_) => url);
```
Comment (coroiu) [inline apps/web/src/app/core/router.service.ts:89]: 🎨  suggestion: I think you can just skip the parameter all together
```suggestion
    await this.deepLinkRedirectUrlState.update(() => url);
```

File: apps/web/src/app/core/router.service.ts [hunk 2]
```diff
@@ -1,24 +1,43 @@
...
 @Injectable()
 export class RouterService {
+  /**
+   * The string value of the URL the user tried to navigate to while unauthenticated.
+   *
+   * Developed to allow users to deep link even when the navigation gets interrupted
+   * by the authentication process.
+   */
+  private deepLinkRedirectUrlState: GlobalState<string>;
+
```
Comment (coroiu) [inline apps/web/src/app/core/router.service.ts:28]: ❓ question: will this be migrated to a GET parameter eventually?

File: apps/web/src/app/core/router.service.ts [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
-import { StateService } from "@bitwarden/common/platform/abstractions/state.service";
 import { Utils } from "@bitwarden/common/platform/misc/utils";
+import {
+  KeyDefinition,
+  ROUTER_DISK,
+  StateProvider,
+  GlobalState,
+} from "@bitwarden/common/platform/state";
+
+export const DEEP_LINK_REDIRECT_URL = new KeyDefinition(ROUTER_DISK, "deepLinkRedirectUrl", {
```
Comment (coroiu) [inline apps/web/src/app/core/router.service.ts:15]: ❌ issue?: does this need to be exported? I haven't worked with state providers so much so I might just be missing something :)

```suggestion
const DEEP_LINK_REDIRECT_URL = new KeyDefinition(ROUTER_DISK, "deepLinkRedirectUrl", {
```

File: apps/web/src/app/core/router.service.ts (hunk 2 above)
Comment (rr-bw) [reply apps/web/src/app/core/router.service.ts:28]: Could you expand a bit on what you mean by that?

Comment (coroiu) [reply apps/web/src/app/core/router.service.ts:28]: Some sites put redirects like this into the URL as a GET parameter, so I was just wondering if this something we might do or if we maybe have a special requirement that needs to use the state

Comment (rr-bw) [reply apps/web/src/app/core/router.service.ts:28]: @ike-kottlowski would you have any background on this? Is there a reason the original URL needs to be stored in state? I believe you worked on the original implementation so just curious if you have any extra insight.

**Review by ike-kottlowski — APPROVED** [review]

**Review by coroiu — APPROVED** [review]

**Pull request merged**